tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -1042,22 +1043,30 @@ REDUCTION_OP(OpType::All_Zero, (std::all_of));
 
 #undef REDUCTION_OP
 
-template <typename T> struct Op<OpType::Dot, T, 2> : DefaultValidation<T> {};
+template <typename T> struct Op<OpType::Dot, T, 2> : StrictValidation {};
 template <typename T> struct ExpectedBuilder<OpType::Dot, T> {
-  static std::vector<T> buildExpected(Op<OpType::Dot, T, 2>,
+  static std::vector<T> buildExpected(Op<OpType::Dot, T, 2> Op,
                                       const InputSets<T> &Inputs,
                                       uint16_t ScalarInputFlags) {
     UNREFERENCED_PARAMETER(ScalarInputFlags);
 
     // Accumulate in fp32 to improve precision.
     float DotProduct = 0.0f;
+    float DotProductAbs = 0.0f;
 
-    for (size_t I = 0; I < Inputs[0].size(); ++I) {
+    const size_t VectorSize = Inputs[0].size();
+
+    for (size_t I = 0; I < VectorSize; ++I) {
       const float A = Inputs[0][I];
       const float B = Inputs[1][I];
       DotProduct += A * B;
+
+      DotProductAbs += std::fabs(A) * std::fabs(B);
     }
 
+    Op.ValidationConfig.Tolerance =
+        (VectorSize + 1) * std::numeric_limits<T>::epsilon() * DotProductAbs;
+
     std::vector<T> Expected;
```

Comment:
I'm a bit confused about what effect this actually has.  `Op` is passed by value into `buildExpected`, wich means that this is setting the tolerance value on a local value that's then discarded when this function returns.